include/af/algorithm.h

```diff
@@ -71,6 +103,39 @@ namespace af
     AFAPI array product(const array &in, const int dim, const double nanval);
 #endif
 
+#if AF_API_VERSION >= 37
+    /**
+       C++ Interface for product of elements in an array according to a key
+
+       \param[out] keys_out will contain the reduced keys in \p vals along \p dim
+       \param[out] vals_out will contain the product of all values in \p vals along \p dim according to \p keys
+       \param[in] keys is the key array
+       \param[in] vals is the array containing the values to be reduced
+       \param[in] dim The dimension along which the add operation occurs
+
+       \ingroup reduce_func_product
+
+       \note \p dim is -1 by default. -1 denotes the first non-singleton dimension.
+
+    */
+    AFAPI void productByKey(array &keys_out, array &vals_out, const array &keys, const array &vals, const int dim = -1);
+
+    /**
+       C++ Interface for product of elements in an array according to a key while replacing nan values
+
+       \param[out] keys_out will contain the reduced keys in \p vals along \p dim
+       \param[out] vals_out will contain the product of all values in \p vals along \p dim according to \p keys
+       \param[in] keys is the key array
+       \param[in] vals is the array containing the values to be reduced
+       \param[in] dim The dimension along which the add operation occurs
+       \param[in] nanval Replace nans with the value passed to this function
```

Comment:
Really minor, but "passed to this function" seems like a typo that should have been "passed to this argument". Some of the other functions that replace nans describe this parameter like this: "nanval is the value which replaces nans"